src/operator/tensor/dot-inl.cuh

```diff
@@ -45,11 +46,51 @@ struct DotCsrDnsDns {
 };
 
 /*!
- * \brief Kernel of dot(csr.T(), dns1) = dns2
- * Parallelization by output matrix elements
+ * \brief Vector kernel of dot(csr, dns1) = dns2
+ * Parallelization by output matrix elements: 1 warp/element
  */
 template<int req>
-struct DotCsrTransDnsDns {
+struct DotCsrDnsDnsVectorKernel {
+  template<typename DType, typename IType, typename CType>
+  __device__ __forceinline__ static void Map(int tid, DType* out, const DType* data_l, const IType* indptr_l,
+                                             const CType* col_idx_l, const DType* data_r,
+                                             const int num_cols_r) {
+    __shared__ volatile DType vals[kBaseThreadNum];
+
+    const int warp_id = tid / 32;           // global warp id
+    const int lane = tid & (32-1);          // local thread id within warp
+    const int irow = warp_id / num_cols_r;  // lhs row that this warp computes
+    const int kcol = warp_id % num_cols_r;  // rhs column that this warp computes
+
+    // Range of nnz elements in this row
+    const int low  = static_cast<int>(indptr_l[irow]);
+    const int high = static_cast<int>(indptr_l[irow+1]);
+
+    // Compute running sum per thread
+    vals[threadIdx.x] = 0;
+    for (int j = low+lane; j < high; j+=32) {
+      vals[threadIdx.x] += data_l[j] * data_r[col_idx_l[j]*num_cols_r + kcol];
+    }
+
+    // Parallel reduction in shared memory
+    if (lane < 16) vals[threadIdx.x] += vals[threadIdx.x+16];
```

Comment:
Generally you may leave it even in CUDA 9, in the reduction part it should not matter from the performance point of view (it is there just to stop compiler from making optimizations that could break the warp-synchronous code).
However, it should already matter in the preprocessing part (Line 72), since the compiler has to write and read from shared memory during every iteration of the loop. Why not use a normal variable (that will reside in a register) for this and write the result to shared only at the end?